Persisting tide jobs.

jeremykross opened the pull request:
@normanjoyner 

Here's some code to persist the tide jobs in myriad.  

I'm a little bit confused about how we're intending to handle the case of jobs that get issued on a leader that's not the controlling leader.  The code here assumes jobs are issued to the controlling leader, but will propagate those jobs to new leaders as they get elected.  

Please clarify if that's not the desired use case.  

fixes https://github.com/containership/containership/issues/173

Depends on: https://github.com/containership/containership.core/pull/77

nicktate commented on lib/tide-scheduler.js:
```diff
@@ -5,18 +5,54 @@ var Job = require([__dirname, "job"].join("/"));
-        var self = this;
+    initialize: function(core) {
         this.core = core;
+        this.myriad_key = [this.core.constants.myriad.VARIABLES_PREFIX, "tide-jobs"].join(this.core.constants.myriad.DELIMITER);
```
I would say stray away from using the `VARIABLES_PREFIX`. I believe we will be using that for cluster wide variables that get displayed to the user on the frontend. It is probably best to just create a new `TIDE-JOB` constant that is just like `containership::tide-jobs` inside the `core.constants` and reference that here.

nicktate commented on lib/tide-scheduler.js:
```diff
@@ -32,58 +68,90 @@ module.exports = {
+    },
+
+    defrostJobs: function(fn) {
+        this.core.cluster.myriad.persistence.get(this.myriad_key, (err, jobs) => {
```
Did you verify `jobs` comes back as a actual json object, I think it might actually come back as just a `string` that you need to parse.

normanjoyner commented on lib/tide-scheduler.js:
```diff
@@ -5,18 +5,54 @@ var Job = require([__dirname, "job"].join("/"));
+            this.jobs[config.id].schedule(() => {
+                this.core.loggers["tide-scheduler"].log("verbose", ["Created tide application", config.application.id].join(" "));
+
+                this.core.applications.remove(config.application.id, (err) => {
```
what is the possible value of `err` here? can we write a debug level message stating that the application does not yet exist, if the error reports that?

normanjoyner commented on lib/tide-scheduler.js:
```diff
@@ -5,18 +5,54 @@ var Job = require([__dirname, "job"].join("/"));
+    schedule_job: function(config) {
+        if(_.has(this.jobs, config.id)) {
+            this.jobs[config.id].schedule(() => {
+                this.core.loggers["tide-scheduler"].log("verbose", ["Created tide application", config.application.id].join(" "));
```
The application should already be created on the cluster. This message might be more useful if it specified that the job was starting `n` number of containers.

normanjoyner commented on lib/tide-scheduler.js:
```diff
@@ -5,18 +5,54 @@ var Job = require([__dirname, "job"].join("/"));
+            this.defrostJobs(on_defrost);
+        }
+
+        this.core.cluster.legiond.on("myriad.bootstrapped", () => {
```
is it necessary to listen for / respond to this event?

normanjoyner replied:
it may be better to change this to `.on('promoted')`

normanjoyner replied:
:+1: `myriad.bootstrapped` is correct here

normanjoyner commented on lib/tide-scheduler.js:
```diff
@@ -32,58 +68,90 @@ module.exports = {
+                if(!_.isEmpty(this.jobs)) {
+                    _.forEach(this.jobs, (j) => {
+                        j.cancel();
+                        this.core.applications.remove(j.config.id, _.identity);
```
since this is async, will this affect functions below?

jeremykross commented:
@normanjoyner  @nicktate  - I updated things as requested.

@normanjoyner  - I modified the logic around the asynchronous job removal to take away some of the ambiguity as to whether or not the code works as intended.

normanjoyner commented on lib/tide-scheduler.js:
```diff
@@ -5,18 +5,55 @@ var Job = require([__dirname, "job"].join("/"));
+    schedule_job: function(config) {
+        if(_.has(this.jobs, config.id)) {
+            this.jobs[config.id].schedule(() => {
+                this.core.loggers["tide-scheduler"].log("verbose", ["Deplyoing tide application", config.application.id, "to", config.instances, "containers."].join(" "));
```
typo on "Deploying"
